GVFS/GVFS/CommandLine/SparseVerb.cs

```diff
@@ -93,6 +93,44 @@ private enum SetDirectoryTimeResult
 
         protected override string VerbName => SparseVerbName;
 
+        internal static bool PathCoveredBySparseFolders(ref int index, string statusOutput, HashSet<string> sparseFolders, out string gitPath)
+        {
+            int endOfPathIndex = statusOutput.IndexOf(StatusPathSeparatorToken, index);
+            gitPath = statusOutput.Substring(index, endOfPathIndex - index);
+            index = endOfPathIndex + 1;
+
+            string filePath = gitPath.Replace(GVFSConstants.GitPathSeparator, Path.DirectorySeparatorChar);
+            if (sparseFolders.Any(x => filePath.StartsWith(x + Path.DirectorySeparatorChar, GVFSPlatform.Instance.Constants.PathComparison)))
+            {
+                // Path is covered by a recursive entry
+                return true;
+            }
+
+            int pathSeparatorIndex = filePath.LastIndexOf(Path.DirectorySeparatorChar);
+            if (pathSeparatorIndex < 0)
+            {
+                // Path is in the root, and root entries are always in the sparse set
+                return true;
+            }
+
+            // Get the parent path (including the path separator)
+            string parentPath = filePath.Substring(startIndex: 0, length: pathSeparatorIndex + 1);
+            if (sparseFolders.Any(x => x.StartsWith(parentPath, GVFSPlatform.Instance.Constants.PathComparison)))

```

Comment:
This has a less-obvious perf change, but creating a `HashSet<string>` full of the "parent paths" (i.e. those that are parents of the sparse folders, but not actually in the sparse folders) we can test `parentFolders.Contains(parentPath)`. [Here is the equivalent Git code](https://github.com/derrickstolee/git/blob/7577ffc034be90bd60da17e8f3189d999e7f8149/dir.c#L631-L717), but the sparse-checkout file already has the full list of parent paths as part of the list, so it may not be incredibly helpful.

The only reason I bring this up is because the current algorithm is quadratic: O(N*M) where N is the number of sparse folders and M is the number of dirty paths. Using the `HashSet<string>` method I describe will be O(N+M*d) (where d is the max depth of a dirty path) because O(N) time to construct the hash sets and O(M*d) time to check the values.

Reply:
Thanks for the great perf suggestions.  I think I will file a follow up for these so that those changes can be isolated as a performance only PR.

Reply:
Filed #1608